### DISCUSSION TITLE
C# script guidance needed (partition definition changes)

### DISCUSSION THREAD

**Discussion by MateuszMo** [TabularEditor/TabularEditor3#1184] [🙋♀️TE3 Q&A]
Hi,

I know there is no better place to ask about C# scripting for tabular models. Therefore I allowed myself to reach out to you as I struggle a bit with my attempts. I tried to create some script that will help me transform a data model which imports the data from "real" sources to a data model that simply queries the copy of the same data model that sits in the premium workspace (which allows to query it like it was AAS database).

I attached the c# code at the very end of this message.

Everything works fine at first glance - two parameters and a function get created, partitions of non calculated tables get overwritten with the intended code but then I fail to actually deploy those changes. Is there anything I am missing? I will be super grateful for any hints.

Regards,

Mateusz 

```
// define workspace connection
string workspace_connection  = "powerbi://api.powerbi.com/v1.0/myorg/00-DataHub-CE-Import%20%5BTest%5D";
// define data model name
string datamodel_name = "Shipments Dataset";

// delete all existing shared expressions
foreach(var exp in Model.Expressions.ToList())
{
  exp.Delete();
}

// create a custom fReadTable function
// it is meant to create a query to pull the data from the premium workspace data model acting as AAS database
// and to remove not needed prefixes and suffixes for column names so that they stay "unchanged"
var CustomFunction = Model.AddExpression("fReadTable");
CustomFunction.Kind = ExpressionKind.M;
CustomFunction.Expression = "(_vTableName as text, _vWorkspaceCon as text, _vDatamodel as text) =>\n\nlet\n    sourceTable = AnalysisServices.Database(\n        _vWorkspaceCon,\n        _vDatamodel,\n        [\n            Query = \"\n            EVALUATE\n            '\" & _vTableName & \"'\n            \"\n        ]\n    ),\n    columnNames = Table.ColumnNames(sourceTable),\n    removeCharacters = (columnName as text) =>\n        Text.Replace(Text.Replace(columnName, _vTableName & \"[\", \"\"), \"]\", \"\"),\n    cleanedColumnNames = List.Transform(columnNames, each removeCharacters(_)),\n    outputTable = Table.RenameColumns(sourceTable, List.Zip({columnNames, cleanedColumnNames}))\nin\n    outputTable";

// create a workspace parameter
var WorskpaceParameter = Model.AddExpression("01_Workspace_Connection");
WorskpaceParameter.Kind = ExpressionKind.M;
WorskpaceParameter.Expression = "\"" + workspace_connection + "\" meta [IsParameterQuery=true, List={\"" + workspace_connection + "\"}, DefaultValue=\"" + workspace_connection + "\", Type=\"Text\", IsParameterQueryRequired=true]";

//create a datamodel parameter
var DatamodelParameter  = Model.AddExpression("02_Datamodel_Name");
DatamodelParameter.Kind = ExpressionKind.M;
DatamodelParameter.Expression = "\"" + datamodel_name + "\" meta [IsParameterQuery=true, List={\"" + datamodel_name + "\"}, DefaultValue=\"" + datamodel_name + "\", Type=\"Text\", IsParameterQueryRequired=true]";

//iterate through all non calculated tables
//overwrite each expression to use the above defined custom fReadTable function
foreach (var t in Model.Tables)
{
    if (Convert.ToString(t.SourceType) != "Calculated")
    {
        foreach (var p in t.Partitions)
        {
            if (p.Expression != "" & p.Name != "Partition")
            {
                p.Expression = "let\r\n    Source = fReadTable(\"" + t.Name + "\", #\"01_Workspace_Connection\", #\"02_Datamodel_Name\")\r\nin\r\n    Source";
            }

        }
    }
}
```

**Comment by otykier** [comment]
Hi @MateuszMo 
From first glance, your script looks fine (assuming you intend to import data from the "Shipments Dataset" model, into the current model, directly using DAX).
Could you provide some more details on where this doesn't work? Does the script fail to run, or are you getting errors later on when attempting to deploy this model? Have you tried to "manually" set up a model that imports from the "Shipments Dataset", and compared the configuration of such a model to the current one after running the script?

**Reply by MateuszMo** [reply]
Hi @otykier,
Thanks a lot for a quick response!
Yes, when set up manually it works like a charm 🙂
Let me paste here a link to the article where I tried to explain it in more details - hope it will clarify where I am stuck and what I was able to capture when it comes to the errors 🙂

Adjusting overall partitions definitions logic through C# script in Tabular Editor challenge https://medium.com/@mossakowski.mateusz/adjusting-overall-partitions-definitions-logic-through-c-script-in-tabular-editor-challenge-2507071c96f7

**Reply by otykier** [reply]
So I'm assuming this is the main problem?

> After running the script and applying the changes to the model, I encountered an issue when attempting to save the changes to the .pbix file. It appears that Power BI is “ignoring” the applied changes and refreshing the data model to its previous state, pointing to the real source data instead of the AAS/premium workspace model.

If so, I'm afraid I have some bad news: Making changes to _M expressions or partitions_ of a model through external tools (Tabular Editor) while connected to the AS instance in Power BI Desktop is **not supported by Microsoft** ([official documentation](https://learn.microsoft.com/en-us/power-bi/developer/projects/projects-overview#model-authoring)). You may be able to workaround this limitation by using Developer Mode (aka. Power BI Desktop Projects) as described in the link. Another solution could be to make the change against a model that has been deployed to the Power BI service (using XMLA endpoint).

**Reply by MateuszMo** [reply]
@otykier, Thanks once more! It appears executing the script in TE against the bim file from the pbip folder resolves everything! I'll confirm today, but it seems the hiccup was running the script against pbix, not pbip.

Your Saint Nicolas Day gift arrived two days early on the 6th of December! 😄
